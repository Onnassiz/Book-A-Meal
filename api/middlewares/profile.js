import isUrl from 'is-url';

const { check } = require('express-validator/check');

export const profileFormConstraints = [
  check('businessName')
    .exists()
    .withMessage('the business name field is require')
    .isLength({ min: 1 })
    .withMessage('the business name field is required')
    .isString()
    .withMessage('the business name must be a string')
    .trim(),

  check('mission')
    .exists()
    .withMessage('the business name field is require')
    .isLength({ min: 1 })
    .withMessage('the business name field is required')
    .isString()
    .withMessage('the business name must be a string')
    .trim(),

  check('contact')
    .exists()
    .withMessage('the contact field is require')
    .isLength({ min: 1 })
    .withMessage('the contact field is required')
    .isString()
    .withMessage('the contact must must be a string')
    .trim(),

  check('email')
    .exists()
    .withMessage('the email field is require')
    .isLength({ min: 1 })
    .withMessage('the email field is required')
    .isEmail()
    .withMessage('the email field must be an email')
    .trim()
    .normalizeEmail(),

  check('banner')
    .optional({ nullable: true })
    .custom((value) => {
      return isUrl(value);
    })
    .withMessage('image link must be a URL'),
];

export const profileBannerConstraints = [
  check('banner')
    .exists()
    .withMessage('the image field is require')
    .isLength({ min: 1 })
    .withMessage('the image field is required')
    .optional({ nullable: true })
    .custom((value) => {
      return isUrl(value);
    })
    .withMessage('image link must be a URL'),
];
