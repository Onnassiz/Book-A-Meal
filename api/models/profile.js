
module.exports = (sequelize, DataTypes) => {
  const Profile = sequelize.define('profile', {
    id: {
      allowNull: false,
      primaryKey: true,
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
    },
    businessName: DataTypes.STRING,
    contact: DataTypes.STRING,
    email: DataTypes.STRING,
    mission: DataTypes.STRING,
    banner: DataTypes.STRING,
  }, {});
  Profile.associate = (models) => {
    Profile.belongsTo(models.user);
    // Profile.belongsTo(models.mealOrder);
  };
  return Profile;
};

