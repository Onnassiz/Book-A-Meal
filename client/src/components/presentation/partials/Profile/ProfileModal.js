import React from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-modal';
import empty from 'is-empty';
import { addProfileModalStyle } from './../../../../utilities/modalStyles';
import { BasicInput, TextArea } from '../../../presentation/form/BasicInput';
import ShowErrors from './../../partials/ShowErrors';
import SubmitButton from '../../../presentation/form/SubmitButton';

const closeModalStyle = {
  float: 'right',
};

const renderInputFields = data => (
  <div>
    <BasicInput
      name="businessName"
      type="text"
      label="Business Name"
      value={data.state.businessName}
      onChange={data.onChange}
      hasError={data.state.errors.businessName !== undefined}
    />
    <BasicInput
      name="mission"
      type="text"
      label="Mission Statement"
      value={data.state.mission}
      onChange={data.onChange}
      hasError={data.state.errors.mission !== undefined}
    />
    <BasicInput
      name="email"
      type="text"
      label="Business Email"
      value={data.state.email}
      onChange={data.onChange}
      hasError={data.state.errors.email !== undefined}
    />
  </div>
);

const renderModalBody = props => (
  <div>
    <div className="col-12">
      <a onClick={props.handleAddButtonClick} style={closeModalStyle}><i style={{ fontSize: 25 }} className="material-icons">close</i></a>
    </div>
    <div className="box">
      <h3>Setup Business Profile</h3>
      <ShowErrors clientErrors={props.state.errors} serverErrors={props.profile.errors || {}} />
      <form onSubmit={props.handleSubmit}>
        {renderInputFields(props)}
        <TextArea
          name="contact"
          type="text"
          label="Contact"
          value={props.state.contact}
          onChange={props.onChange}
          hasError={props.state.errors.contact !== undefined}
        />
        <SubmitButton value={empty(props.profile.businessName) ? 'Submit' : 'Update'} isLoading={props.formState.isLoading} />
      </form>
    </div>
  </div>
);

const ProfileModal = props => (
  <div>
    <Modal
      isOpen={props.state.isShowingModal}
      closeTimeoutMS={1}
      style={addProfileModalStyle}
      ariaHideApp={false}
      contentLabel="Modal"
      className="create-profile"
    >
      {renderModalBody(props)}
    </Modal>
  </div>
);

ProfileModal.propTypes = {
  state: PropTypes.object.isRequired,
};

renderModalBody.propTypes = {
  state: PropTypes.object.isRequired,
  profile: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  formState: PropTypes.func.isRequired,
  handleAddButtonClick: PropTypes.func.isRequired,
};

export default ProfileModal;
