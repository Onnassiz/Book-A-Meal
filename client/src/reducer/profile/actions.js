import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actions';

export const SET_PROFILE = 'SET_PROFILE';
export const SET_PROFILE_SERVER_ERRORS = 'SET_PROFILE_SERVER_ERRORS';
export const SET_PROFILE_ALERT = 'SET_PROFILE_ALERT';

function setAlert(alert) {
  return dispatch => dispatch({
    type: SET_PROFILE_ALERT,
    alert,
  });
}

export function setProfile(profile) {
  return dispatch => dispatch({
    type: SET_PROFILE,
    id: profile.id,
    businessName: profile.businessName,
    mission: profile.mission,
    contact: profile.contact,
    banner: profile.banner,
    email: profile.email,
  });
}

function setProfileErrors(errors) {
  return dispatch => dispatch({
    type: SET_PROFILE_SERVER_ERRORS,
    errors,
  });
}

export function postProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.post('api/v1/profile', profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully created'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function updateProfile(profile) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Profile successfully updated'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function putImage(id, field) {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.put(`api/v1/profile/image/${id}`, field).then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      dispatch(setAlert('Banner successfully uploaded'));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      dispatch(setProfileErrors(error.response.data));
      return error;
    });
  };
}

export function getProfile() {
  return (dispatch) => {
    dispatch(setLoading());
    return axios.get('api/v1/profile').then((response) => {
      dispatch(unsetLoading());
      dispatch(setProfile(response.data));
      return response;
    }).catch((error) => {
      dispatch(unsetLoading());
      return error;
    });
  };
}
