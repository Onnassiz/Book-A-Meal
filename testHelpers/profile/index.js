import { profile } from '../../api/models';

import { getCatererId } from '../main';

const deleteProfile = (done) => {
  profile.destroy({
    where: {},
  }).then(() => {
    done();
  });
};

const insertProfile = done => getCatererId(done, false).then((id) => {
  const newProfile = profile.build({
    businessName: 'Your name',
    mission: 'Your mission',
    contact: 'Your this is your contact',
    banner: 'https://banner.com',
    email: 'benjamin@gmail.com',
    userId: id,
  });

  return newProfile.save().then(data => data);
});

export {
  deleteProfile,
  insertProfile,
};
