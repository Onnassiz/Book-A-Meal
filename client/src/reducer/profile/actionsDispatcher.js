import axios from 'axios';
import { setLoading, unsetLoading } from '../formState/actionsDispatcher';
import { setProfileAction, setProfileErrorsAction } from './actions';


export const setProfile = profile => dispatch => dispatch(setProfileAction(profile));

const setProfileErrors = errors => dispatch => dispatch(setProfileErrorsAction(errors));

export const postProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.post('api/v1/profile', profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const updateProfile = profile => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${profile.id}`, profile).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const putImage = (id, field) => (dispatch) => {
  dispatch(setLoading());
  return axios.put(`api/v1/profile/${id}`, field).then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};

export const getProfile = () => (dispatch) => {
  dispatch(setLoading());
  return axios.get('api/v1/profile').then((response) => {
    dispatch(unsetLoading());
    dispatch(setProfile(response.data));
    return response;
  }).catch((error) => {
    dispatch(unsetLoading());
    dispatch(setProfileErrors(error.response.data));
    return error;
  });
};
