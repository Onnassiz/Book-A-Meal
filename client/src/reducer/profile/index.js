import { SET_PROFILE, SET_PROFILE_SERVER_ERRORS } from './actions';

const initialState = {};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_PROFILE:
      return {
        id: action.id,
        businessName: action.businessName,
        email: action.email,
        mission: action.mission,
        contact: action.contact,
        banner: action.banner,
      };
    case SET_PROFILE_SERVER_ERRORS:
      return { ...state, errors: action.errors };
    default:
      return state;
  }
};
