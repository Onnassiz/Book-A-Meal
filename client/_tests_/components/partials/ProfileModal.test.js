import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ProfileModal from '../../../src/components/presentation/partials/Profile/ProfileModal';
import profile from '../../objectProps/profile.props';

const state = profile;
profile.errors = {};

const props = {
  state,
  profile,
  handleSubmit: jest.fn(),
  onChange: jest.fn(),
  formState: { isLoading: true },
  handleAddButtonClick: jest.fn(),
};

const wrapper = shallow(<ProfileModal {...props} />);

describe('<ProfileModal />', () => {
  it('Should render Profile Modal component successfully', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
