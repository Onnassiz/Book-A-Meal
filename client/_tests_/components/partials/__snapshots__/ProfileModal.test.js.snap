// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileModal /> Should render Profile Modal component successfully 1`] = `
<div
  id="profile_modal"
>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    className="create-profile"
    closeTimeoutMS={1}
    contentLabel="Modal"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "WebkitOverflowScrolling": "touch",
          "background": "#ffffff",
          "borderRadius": "6px",
          "bottom": "15%",
          "left": "33%",
          "outline": "none",
          "overflow": "auto",
          "padding": "12px",
          "position": "absolute",
          "right": "34%",
          "top": "10%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.4)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
  >
    <div>
      <div
        className="col-12"
      >
        <a
          onClick={[MockFunction]}
          style={
            Object {
              "float": "right",
            }
          }
        >
          <i
            className="material-icons"
            style={
              Object {
                "fontSize": 25,
              }
            }
          >
            close
          </i>
        </a>
      </div>
      <div
        className="box"
      >
        <h3>
          Setup Business Profile
        </h3>
        <ShowErrors
          clientErrors={Object {}}
          serverErrors={Object {}}
        />
        <form
          onSubmit={[MockFunction]}
        >
          <div>
            <BasicInput
              hasError={false}
              label="Business Name"
              min=""
              name="businessName"
              onChange={[MockFunction]}
              type="text"
              value="The blast"
            />
            <BasicInput
              hasError={false}
              label="Mission Statement"
              min=""
              name="mission"
              onChange={[MockFunction]}
              type="text"
              value="Feeding the world"
            />
            <BasicInput
              hasError={false}
              label="Business Email"
              min=""
              name="email"
              onChange={[MockFunction]}
              type="text"
              value="onnassiz@andela.com"
            />
          </div>
          <TextArea
            hasError={false}
            label="Contact"
            name="contact"
            onChange={[MockFunction]}
            type="text"
            value="12 Benny Road Lagos"
          />
          <SubmitButton
            isLoading={true}
            value="Update"
          />
        </form>
      </div>
    </div>
  </Modal>
</div>
`;
